### PR TITLE
Dala 2403 manual maintenance sprint 51

### PR DESCRIPTION
**Pull request by HussainRoker**
# Pull Request \<manual maintenance sprint>

- manual maintenance sprint (see manual_maintenance.md)
- installed and tested new plugin **Gradle refreshVersions** see https://splitties.github.io/refreshVersions/
- new file added **versions.properties**  this is the only place where dependencies/plugins are to be placed thus providing a centeralized location for updates, also the plugin automatically checks for updates, see link above
- new file added **libs.versions.toml** this enables [version catalog](https://docs.gradle.org/current/userguide/platforms.html#sub:version-catalog)
- pdfbox updated to 3.xx, made relevant changes in code, replaced deprecated api
- updated DockerfileGradle & DockerfileBase to include versions.properties

## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [ ] A peer-review has been executed
  - [ ] The code has been manually inspected by someone who did not implement the feature
- [ ] The PR actually implements what is described in the JIRA-Issue
- [ ] At least one test exists testing the new feature
  - [ ] If you have created new test files, make sure that they are included in a test container and actually run in the CI
- [ ] Documentation is updated as required
- [ ] The automated deployment is updated if required
- [ ] If there was a database entity class added, there must also be a migration script for creating the corresponding database if flyway is already used by the service
- [ ] IF there are changes done to the framework data models or to a database entity class, the following steps were completed
  - [ ] The main branch is deployed to a dev server with `Reset non-user related Docker Volumes & Re-populate` turned on
  - [ ] It's verified that the CD run is green
  - [ ] Immediately after that the feature branch is deployed to the same server with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [ ] It's verified that the CD run is green
  - [ ] The new feature is manually used/tested/observed on the dev server
- [ ] ELSE, the new version is deployed to the dev server "dev1" using this branch
  - [ ] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on 
  - [ ] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [ ] It's verified that the CD run is green
  - [ ] The new feature is manually used/tested/observed on dev server
  - [ ] All implemented Social Logins have been tested manually in the UI
- [ ] Make sure that pull requests in the development tools repository, which are connected to this pull requests, 
are merged

### CODE REVIEW

File: dataland-keycloak/realms/master-realm.json [hunk 1]
```diff
@@ -2393,7 +2393,7 @@
     "frontendUrl": "http://dataland-admin:6789/keycloak",
     "acr.loa.map": "{}"
   },
-  "keycloakVersion": "22.0.1",
+  "keycloakVersion": "22.0.3",
```
Comment (fsoffel) [inline dataland-keycloak/realms/master-realm.json:2396]: The container keycloak version, the version in the other realm file and the keycloak-js version are still on 22.0.1. Is this intended?

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 2]
```diff
@@ -11,7 +11,7 @@ creation URL (or simply copy this md file into the description)
 
 The following known issues need to be reviewed in case a compatible version is available. Add new known issues as they
 appear.
-
+- [ ] Update node.js to version 20.6.1 breaks the build
```
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: There is no update on any node package at all?

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -2,5 +2,6 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.1-bin.zip
 networkTimeout=10000
+validateDistributionUrl=true
```
Comment (fsoffel) [inline gradle/wrapper/gradle-wrapper.properties:5]: Why was this added? Were there problems?

File: versions.properties [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+####
+#### Don't manually edit or split the comments that start with four hashtags (####),
+#### they will be overwritten by refreshVersions.
+####
+#### suppress inspection "SpellCheckingInspection" for whole file
+#### suppress inspection "UnusedProperty" for whole file
+####
+#### NOTE: Some versions are filtered by the rejectVersionIf predicate. See the settings.gradle.kts file.
+
+## unused
```
Comment (fsoffel) [inline versions.properties:12]: Is the overall package not used? If so, remove it.

File: versions.properties [hunk 5]
```diff
@@ -0,0 +1,137 @@
+#### Dependencies and Plugin versions with their available updates.
+#### Generated by `./gradlew refreshVersions` version 0.60.2
```
Comment (fsoffel) [inline versions.properties:2]: If I remove this file, it is not regenerated by refreshVersions.

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -0,0 +1,168 @@
+## Generated by $ ./gradlew refreshVersionsCatalog
```
Comment (fsoffel) [inline gradle/libs.versions.toml:1]: If I remove this file, it is not regenerated by refreshVersions.

File: gradle/wrapper/gradle-wrapper.jar
Comment (fsoffel) [inline gradle/wrapper/gradle-wrapper.jar:1]: We usually don't want and have binary files in the git repository. Is this a candidate for .gitignore?

Comment (fsoffel) [reply gradle/wrapper/gradle-wrapper.jar:1]: Hm, apparently this file already existed before this branch. So no need ot address this here.

File: gradlew [hunk 7]
```diff
@@ -130,10 +130,13 @@ location of your Java installation."
     fi
 else
     JAVACMD=java
-    which java >/dev/null 2>&1 || die "ERROR: JAVA_HOME is not set and no 'java' command could be found in your PATH.
+    if ! command -v java >/dev/null 2>&1
```
Comment (fsoffel) [inline gradlew:133]: Why was this changed?

File: dataland-document-manager/src/main/kotlin/org/dataland/documentmanager/services/PdfVerificationService.kt [hunk 8]
```diff
@@ -37,7 +37,7 @@ class PdfVerificationService {
     }
 
     private fun checkIfPdfDocumentIsEmpty(blob: ByteArray, correlationId: String) {
-        PDDocument.load(blob).use {
+        Loader.loadPDF(blob).use {
```
Comment (fsoffel) [inline dataland-document-manager/src/main/kotlin/org/dataland/documentmanager/services/PdfVerificationService.kt:40]: Why was this changed?

File: gradle/libs.versions.toml (hunk 6 above)
Comment (fsoffel) [reply gradle/libs.versions.toml:1]: Ah, that is because it should not be generated entirely by this script.

File: base-dockerfiles/DockerfileGradle [hunk 9]
```diff
@@ -5,5 +5,5 @@ WORKDIR /app
 COPY ./gradle/ ./gradle/
 COPY ./gradlew .
 RUN  ./gradlew --version
-COPY ./build.gradle.kts ./gradle.properties ./settings.gradle.kts .
+COPY ./build.gradle.kts ./gradle.properties ./settings.gradle.kts ./versions.properties ./
```
Comment (fsoffel) [inline base-dockerfiles/DockerfileGradle:8]: Why add the `/`?

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:1]: The gradle versions update procedure must be adapted.

Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:1]: How to get the new version suggestions and how to apply them (uncomment).

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md [hunk 10]
```diff
@@ -23,7 +23,7 @@ appear.
   https://github.com/JLLeitschuh/ktlint-gradle/pull/667)
 - [ ] Update @zxcvbn-ts/language-common to 3.0.3 is skipped due to issues in rebuilding keycloak Docker images
 - [ ] Update @zxcvbn-ts/language-en to 3.0.1 is skipped due to issues in rebuilding keycloak Docker images
-- [ ] Flyway version from 9.18.0 and above (until 9.21.1) seem to cause issues. Future versions should be tried
+- [ ] Flyway version from 9.18.0 and above (until 9.22.1) seem to cause issues. Future versions should be tried
```
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:26]: But we migrated to this version. So either remove this line or revert the version. We should definitely specify the issues we have more clearly here.

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:1]: ktlint got updated. If this works, you can remove the corresponding line here.

File: gradlew (hunk 7 above)
Comment (marcelrasbachdfine) [reply gradlew:133]: this change was done by running the gradle wrapper

File: gradle/wrapper/gradle-wrapper.properties (hunk 3 above)
Comment (marcelrasbachdfine) [reply gradle/wrapper/gradle-wrapper.properties:5]: no this is a result of running the gradle wrapper
(shouldn't this be the case when running it?)

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 10 above)
Comment (marcelrasbachdfine) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md]: line removed

File: dataland-keycloak/realms/master-realm.json (hunk 1 above)
Comment (marcelrasbachdfine) [reply dataland-keycloak/realms/master-realm.json:2396]: done

File: dataland-frontend/package.json [hunk 11]
```diff
@@ -59,6 +59,7 @@
     "mocha": "^10.2.0",
     "npm-check-updates": "^16.14.4",
     "otplib": "^12.0.1",
+    "sass": "^1.67.0",
```
Comment (fsoffel) [inline dataland-frontend/package.json:62]: Why was this added?

File: gradlew (hunk 7 above)
Comment (fsoffel) [reply gradlew:133]: 👍

File: gradle/wrapper/gradle-wrapper.properties (hunk 3 above)
Comment (fsoffel) [reply gradle/wrapper/gradle-wrapper.properties:5]: I don't know. I guess this is perfectly fine. I just never observed something like this before.

File: dataland-keycloak/realms/master-realm.json (hunk 1 above)
Comment (fsoffel) [reply dataland-keycloak/realms/master-realm.json:2396]: Keycloak dockerfile line 14 is missing the update. This is the crucial point.

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 2 above)
Comment (fsoffel) [inline .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: This got updated to 20.6.2

Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: In keycloak, my bad

File: dataland-frontend/package.json (hunk 11 above)
Comment (HussainRoker) [reply dataland-frontend/package.json:62]: @fsoffel you mean the node version? only @types/node was updated not node itself, I updated because `updatepackages` suggested it and nothing broke,  I can downgrade to previous if needed.

Comment (HussainRoker) [reply dataland-frontend/package.json:62]: @fsoffel sass was required dependency for vite at least for the latest version

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 2 above)
Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: @fsoffel only @types/node was updated not node itself, I updated because updatepackages suggested it and nothing broke, I can downgrade to previous if needed.

Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: Where would we update node?

Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: dockerFile

Comment (HussainRoker) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: @fsoffel see docker-compose.yml (line 203)

File: dataland-frontend/package.json (hunk 11 above)
Comment (fsoffel) [reply dataland-frontend/package.json:62]: Dont packages manage that by themselves in the package-lock.json?

File: .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md (hunk 2 above)
Comment (fsoffel) [reply .github/PULL_REQUEST_TEMPLATE/manual_maintenance.md:14]: Ah, yeah. Thanks. The question stays, should we update @types/node that. I don't really know what it is responsible for. But as long as nothing seems to break, this is fine.
